<a href="https://colab.research.google.com/github/popkdodge/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/module3-join-and-reshape-data/LS_DS_113_Join_and_Reshape_Data_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 1, Sprint 1, Module 3*

---

# Join and Reshape datasets

Objectives
- concatenate data with pandas
- merge data with pandas
-  understand tidy data formatting
-  melt and pivot data with pandas

Links
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
- [Tidy Data](https://en.wikipedia.org/wiki/Tidy_data)
  - Combine Data Sets: Standard Joins
  - Tidy Data
  - Reshaping Data
- Python Data Science Handbook
  - [Chapter 3.6](https://jakevdp.github.io/PythonDataScienceHandbook/03.06-concat-and-append.html), Combining Datasets: Concat and Append
  - [Chapter 3.7](https://jakevdp.github.io/PythonDataScienceHandbook/03.07-merge-and-join.html), Combining Datasets: Merge and Join
  - [Chapter 3.8](https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html), Aggregation and Grouping
  - [Chapter 3.9](https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html), Pivot Tables
  
Reference
- Pandas Documentation: [Reshaping and Pivot Tables](https://pandas.pydata.org/pandas-docs/stable/reshaping.html)
- Modern Pandas, Part 5: [Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [3]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

--2020-05-07 14:42:44--  https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.8.237
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.8.237|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 205548478 (196M) [application/x-gzip]
Saving to: ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’

instacart_online_gr 100%[===================>] 196.03M  38.3MB/s    in 4.9s    

2020-05-07 14:42:49 (40.0 MB/s) - ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’ saved [205548478/205548478]



In [4]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

instacart_2017_05_01/
instacart_2017_05_01/._aisles.csv
instacart_2017_05_01/aisles.csv
instacart_2017_05_01/._departments.csv
instacart_2017_05_01/departments.csv
instacart_2017_05_01/._order_products__prior.csv
instacart_2017_05_01/order_products__prior.csv
instacart_2017_05_01/._order_products__train.csv
instacart_2017_05_01/order_products__train.csv
instacart_2017_05_01/._orders.csv
instacart_2017_05_01/orders.csv
instacart_2017_05_01/._products.csv
instacart_2017_05_01/products.csv


In [5]:
%cd instacart_2017_05_01

/content/instacart_2017_05_01


In [6]:
!ls -lh *.csv

-rw-r--r-- 1 502 staff 2.6K May  2  2017 aisles.csv
-rw-r--r-- 1 502 staff  270 May  2  2017 departments.csv
-rw-r--r-- 1 502 staff 551M May  2  2017 order_products__prior.csv
-rw-r--r-- 1 502 staff  24M May  2  2017 order_products__train.csv
-rw-r--r-- 1 502 staff 104M May  2  2017 orders.csv
-rw-r--r-- 1 502 staff 2.1M May  2  2017 products.csv


# Assignment

## Join Data Practice

These are the top 10 most frequently ordered products. How many times was each ordered? 

1. Banana
2. Bag of Organic Bananas
3. Organic Strawberries
4. Organic Baby Spinach 
5. Organic Hass Avocado
6. Organic Avocado
7. Large Lemon 
8. Strawberries
9. Limes 
10. Organic Whole Milk

First, write down which columns you need and which dataframes have them.

Next, merge these into a single dataframe.

Then, use pandas functions from the previous lesson to get the counts of the top 10 most frequently ordered products.

In [7]:
##### YOUR CODE HERE #####
"""  I would need the product name column and merge with the order dataframe.  
Then I would try to find the values counts of the dataframe.             """

'  I would need the product name column and merge with the order dataframe.  \nThen I would try to find the values counts of the dataframe.             '

In [0]:
import pandas as pd

In [9]:
orders = pd.read_csv('orders.csv')
print (orders.shape)
orders.head()


(3421083, 7)


order_id  user_id  ... order_hour_of_day  days_since_prior_order
0   2539329        1  ...                 8                     NaN
1   2398795        1  ...                 7                    15.0
2    473747        1  ...                12                    21.0
3   2254736        1  ...                 7                    29.0
4    431534        1  ...                15                    28.0

[5 rows x 7 columns]

In [10]:
products = pd.read_csv('products.csv')
print (products.shape)
products.head()

(49688, 4)


product_id  ... department_id
0           1  ...            19
1           2  ...            13
2           3  ...             7
3           4  ...             1
4           5  ...            13

[5 rows x 4 columns]

In [0]:
order_products__prior = pd.read_csv('order_products__prior.csv')
order_products__train = pd.read_csv('order_products__train.csv')

In [12]:
order_products = pd.concat([order_products__prior, order_products__train])
order_products.head()

order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [0]:
"Now i should merge order_product with order using the reference order id"
order_and_products = pd.merge(order_products, orders, on='order_id', how='inner')


In [14]:
"checking post merge"
"i think I should now merge product to product id"
order_and_products

order_id  product_id  ...  order_hour_of_day  days_since_prior_order
0                2       33120  ...                  9                     8.0
1                2       28985  ...                  9                     8.0
2                2        9327  ...                  9                     8.0
3                2       45918  ...                  9                     8.0
4                2       30035  ...                  9                     8.0
...            ...         ...  ...                ...                     ...
33819101   3421063       14233  ...                 10                     4.0
33819102   3421063       35548  ...                 10                     4.0
33819103   3421070       35951  ...                 10                     8.0
33819104   3421070       16953  ...                 10                     8.0
33819105   3421070        4724  ...                 10                     8.0

[33819106 rows x 10 columns]

In [0]:
order_products_merge_with_product_name = pd.merge(order_and_products, products, on='product_id',how='inner')

In [16]:
order_products_merge_with_product_name.head()

order_id  product_id  ...  aisle_id  department_id
0         2       33120  ...        86             16
1        26       33120  ...        86             16
2       120       33120  ...        86             16
3       327       33120  ...        86             16
4       390       33120  ...        86             16

[5 rows x 13 columns]

In [0]:
"now i am going to omit rows I dont need for the sake of simplicity"
clean_order_products = order_products_merge_with_product_name[['order_id',
                                                               'user_id',
                                                               'product_name']]

In [19]:
clean_order_products['product_name'].value_counts().head(20)
#"""BELOW IS THE ANSWER OF HOW MANY TIMES WAS EACH OF THE TOP 10 is ordered. """

Banana                      491291
Bag of Organic Bananas      394930
Organic Strawberries        275577
Organic Baby Spinach        251705
Organic Hass Avocado        220877
Organic Avocado             184224
Large Lemon                 160792
Strawberries                149445
Limes                       146660
Organic Whole Milk          142813
Organic Raspberries         142603
Organic Yellow Onion        117716
Organic Garlic              113936
Organic Zucchini            109412
Organic Blueberries         105026
Cucumber Kirby               99728
Organic Fuji Apple           92889
Organic Lemon                91251
Organic Grape Tomatoes       88078
Apple Honeycrisp Organic     87272
Name: product_name, dtype: int64



1.   Banana 491,291
2.   Bag of Organic Bananas 394,930
3.   Organic Strawberries 275,577
4.   Organic Baby Spinach 251,705
5.   Organic Hass Avocado 220,977
6.   Organic Avocado  184,224
7.   Large Lemon   160,792
8.   Strawberries  149,445
9.   Limes  146,660
10.  Organic Whole Milk 142,813



## Reshape Data Section

- Replicate the lesson code. Remember, if you haven't followed along typing out what we did during lecture, do that now to get more repetition with the syntax.
- Take table 2 (the transpose of table 1) and practice changing it into Tidy format and back again. You should not use the transpose operation anywhere in this code:
 - Table 2 --> Tidy
 - Tidy --> Table 2
- Load seaborn's `flights` dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.

In [20]:
import pandas as pd
import numpy as np

table1 = pd.DataFrame(
    [[np.nan, 2],
     [16,    11], 
     [3,      1]],
    index=['John Smith', 'Jane Doe', 'Mary Johnson'], 
    columns=['treatmenta', 'treatmentb'])

table2 = table1.T

table2

John Smith  Jane Doe  Mary Johnson
treatmenta         NaN      16.0           3.0
treatmentb         2.0      11.0           1.0

In [0]:
# table2 -> tidy
table2 = table2.reset_index()

In [22]:
table2

index  John Smith  Jane Doe  Mary Johnson
0  treatmenta         NaN      16.0           3.0
1  treatmentb         2.0      11.0           1.0

In [0]:
tidy = table2.melt(id_vars='index',value_vars=['John Smith','Jane Doe','Mary Johnson'])

In [24]:
tidy
#now i need to rename columns

index      variable  value
0  treatmenta    John Smith    NaN
1  treatmentb    John Smith    2.0
2  treatmenta      Jane Doe   16.0
3  treatmentb      Jane Doe   11.0
4  treatmenta  Mary Johnson    3.0
5  treatmentb  Mary Johnson    1.0

In [0]:
tidy = tidy.rename(columns={
    'index':'Treatment',
    'variable':'Name',
    'value':'Result'
})

In [26]:
tidy

Treatment          Name  Result
0  treatmenta    John Smith     NaN
1  treatmentb    John Smith     2.0
2  treatmenta      Jane Doe    16.0
3  treatmentb      Jane Doe    11.0
4  treatmenta  Mary Johnson     3.0
5  treatmentb  Mary Johnson     1.0

In [0]:
# tidy -> table2

In [0]:
expand = tidy.pivot_table(
    index='Treatment',
    columns='Name',
    values='Result'
)

In [29]:
expand

Name        Jane Doe  John Smith  Mary Johnson
Treatment                                     
treatmenta      16.0         NaN           3.0
treatmentb      11.0         2.0           1.0

- Load seaborn's `flights` dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.

In [30]:
import seaborn as sns
flights = sns.load_dataset('flights')


/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [31]:
flights

year      month  passengers
0    1949    January         112
1    1949   February         118
2    1949      March         132
3    1949      April         129
4    1949        May         121
..    ...        ...         ...
139  1960     August         606
140  1960  September         508
141  1960    October         461
142  1960   November         390
143  1960   December         432

[144 rows x 3 columns]

In [0]:
flights_pivot = flights.pivot_table(
    index='year',
    columns='month',
    values='passengers'
)

In [0]:
# Flights Pivot Table

In [34]:
flights_pivot

month  January  February  March  April  ...  September  October  November  December
year                                    ...                                        
1949       112       118    132    129  ...        136      119       104       118
1950       115       126    141    135  ...        158      133       114       140
1951       145       150    178    163  ...        184      162       146       166
1952       171       180    193    181  ...        209      191       172       194
1953       196       196    236    235  ...        237      211       180       201
1954       204       188    235    227  ...        259      229       203       229
1955       242       233    267    269  ...        312      274       237       278
1956       284       277    317    313  ...        355      306       271       306
1957       315       301    356    348  ...        404      347       305       336
1958       340       318    362    348  ...        404      359       310       337
1959       360       342    406    396  ...        463      407       362       405
1960       417       391    419    461  ...        508      461       390       432

[12 rows x 12 columns]

## Join Data Stretch Challenge

The [Instacart blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2) has a visualization of "**Popular products** purchased earliest in the day (green) and latest in the day (red)." 

The post says,

> "We can also see the time of day that users purchase specific products.

> Healthier snacks and staples tend to be purchased earlier in the day, whereas ice cream (especially Half Baked and The Tonight Dough) are far more popular when customers are ordering in the evening.

> **In fact, of the top 25 latest ordered products, the first 24 are ice cream! The last one, of course, is a frozen pizza.**"

Your challenge is to reproduce the list of the top 25 latest ordered popular products.

We'll define "popular products" as products with more than 2,900 orders.



In [0]:
stretch_condition= order_products_merge_with_product_name['order_hour_of_day'] > 14

In [36]:
##### YOUR CODE HERE #####
order_products_merge_with_product_name.head()

order_id  product_id  ...  aisle_id  department_id
0         2       33120  ...        86             16
1        26       33120  ...        86             16
2       120       33120  ...        86             16
3       327       33120  ...        86             16
4       390       33120  ...        86             16

[5 rows x 13 columns]

In [39]:
order_products_merge_with_product_name.shape

(33819106, 13)

In [58]:
small_order_product = order_products_merge_with_product_name[['order_id','product_id','order_hour_of_day','product_name','product_count_total']]

KeyError: ignored

In [57]:
small_order_product

order_id  ...  product_count_total_normalize
0                2  ...                   5.983009e-04
1               26  ...                   5.983009e-04
2              120  ...                   5.983009e-04
3              327  ...                   5.983009e-04
4              390  ...                   5.983009e-04
...            ...  ...                            ...
33819101   2269288  ...                   5.913817e-08
33819102   2507248  ...                   5.913817e-08
33819103   2271346  ...                   2.956908e-08
33819104   3004244  ...                   2.956908e-08
33819105   3376312  ...                   2.956908e-08

[33819106 rows x 6 columns]

In [54]:
small_order_product['product_count_total']= small_order_product.product_name.map(small_order_product.product_name.value_counts())

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [56]:
small_order_product['product_count_total_normalize']= small_order_product.product_name.map(small_order_product.product_name.value_counts(normalize=True))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [49]:
order_products_merge_with_product_name.product_name.value_counts(normalize=True).sort_index().head(10)

#2 Coffee Filters                                2.362570e-05
#2 Cone White Coffee Filters                     4.139672e-07
#2 Mechanical Pencils                            2.661218e-07
#4 Natural Brown Coffee Filters                  7.303564e-06
& Go! Hazelnut Spread + Pretzel Sticks           5.145021e-06
'Swingtop' Premium Lager                         2.956908e-08
(70% Juice!) Mountain Raspberry Juice Squeeze    8.279344e-07
+Energy Black Cherry Vegetable & Fruit Juice     8.870725e-07
.5\" Waterproof Tape                             2.365527e-07
0 Calorie Acai Raspberry Water Beverage          2.927339e-06
Name: product_name, dtype: float64

In [60]:
small_order_product_condition = small_order_product.product_count_total > 2900 & small_order_product.order_hour_of_day > 14

ValueError: ignored

In [0]:
later_day_order = order_products_merge_with_product_name[stretch_condition]

In [25]:
later_day_order.shape

(10796904, 13)

In [46]:
later_day_order.order_hour_of_day.value_counts(normalize=True).head(25)

14    0.171517
15    0.169675
16    0.161558
17    0.133411
18    0.104612
19    0.080396
20    0.062114
21    0.050717
22    0.040397
23    0.025602
Name: order_hour_of_day, dtype: float64

## Reshape Data Stretch Challenge

_Try whatever sounds most interesting to you!_

- Replicate more of Instacart's visualization showing "Hour of Day Ordered" vs "Percent of Orders by Product"
- Replicate parts of the other visualization from [Instacart's blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2), showing "Number of Purchases" vs "Percent Reorder Purchases"
- Get the most recent order for each user in Instacart's dataset. This is a useful baseline when [predicting a user's next order](https://www.kaggle.com/c/instacart-market-basket-analysis)
- Replicate parts of the blog post linked at the top of this notebook: [Modern Pandas, Part 5: Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
##### YOUR CODE HERE #####